api/client/write.go

```diff
@@ -0,0 +1,169 @@
+	option := &pb.WriteTransactionRequest_ModelFindRequest{ModelFindRequest: innerReq}
+	if err = t.client.Send(&pb.WriteTransactionRequest{Option: option}); err != nil {
+		return [][]byte{}, err
+	}
+	var resp *pb.WriteTransactionReply
+	if resp, err = t.client.Recv(); err != nil {
+		return [][]byte{}, err
+	}
+	switch x := resp.GetOption().(type) {
+	case *pb.WriteTransactionReply_ModelFindReply:
+		return x.ModelFindReply.GetEntities(), nil
+	default:
+		return [][]byte{}, fmt.Errorf("WriteTransactionReply.Option has unexpected type %T", x)
+	}
+}
+
+// Create creates new instances of model objects
+func (t *WriteTransaction) Create(items ...interface{}) error {
+	values, err := marshalItems(items)
+	if err != nil {
+		return err
+	}
+	innerReq := &pb.ModelCreateRequest{
+		Values: values,
+	}
+	option := &pb.WriteTransactionRequest_ModelCreateRequest{ModelCreateRequest: innerReq}
+	if err = t.client.Send(&pb.WriteTransactionRequest{Option: option}); err != nil {
+		return err
+	}
+	var resp *pb.WriteTransactionReply
+	if resp, err = t.client.Recv(); err != nil {
+		return err
+	}
+	switch x := resp.GetOption().(type) {
+	case *pb.WriteTransactionReply_ModelCreateReply:
+		for i, entity := range x.ModelCreateReply.GetEntities() {
+			err := json.Unmarshal([]byte(entity), items[i])
+			if err != nil {
+				return err
+			}
+		}
+		return nil
+	default:
+		return fmt.Errorf("WriteTransactionReply.Option has unexpected type %T", x)
+	}
+}
+
+// Save saves existing instances
+func (t *WriteTransaction) Save(items ...interface{}) error {
+	values, err := marshalItems(items)
+	if err != nil {
+		return err
+	}
+	innerReq := &pb.ModelSaveRequest{
+		Values: values,
+	}
+	option := &pb.WriteTransactionRequest_ModelSaveRequest{ModelSaveRequest: innerReq}
+	if err = t.client.Send(&pb.WriteTransactionRequest{Option: option}); err != nil {
+		return err
+	}
+	var resp *pb.WriteTransactionReply
+	if resp, err = t.client.Recv(); err != nil {
+		return err
+	}
+	switch x := resp.GetOption().(type) {
+	case *pb.WriteTransactionReply_ModelSaveReply:
+		return nil
+	default:
+		return fmt.Errorf("WriteTransactionReply.Option has unexpected type %T", x)
+	}
+}
+
+// Delete deletes data
+func (t *WriteTransaction) Delete(entityIDs ...string) error {
+	innerReq := &pb.ModelDeleteRequest{
+		EntityIDs: entityIDs,
+	}
+	option := &pb.WriteTransactionRequest_ModelDeleteRequest{ModelDeleteRequest: innerReq}
+	if err := t.client.Send(&pb.WriteTransactionRequest{Option: option}); err != nil {
+		return err
+	}
+	var resp *pb.WriteTransactionReply
+	var err error
+	if resp, err = t.client.Recv(); err != nil {
+		return err
+	}
+	switch x := resp.GetOption().(type) {
+	case *pb.WriteTransactionReply_ModelDeleteReply:
+		return nil
+	default:
+		return fmt.Errorf("WriteTransactionReply.Option has unexpected type %T", x)
+	}
+}
+
+// End ends the active transaction
+func (t *WriteTransaction) End() error {
+	return t.client.CloseSend()
+}
```

Comment:
Lots of similar code between the read and write transaction files, but it's just different enough that it's not clear to me how to consolidate. If you have any ideas, please let me know.